packages/stateManager/index.ts

```diff
@@ -1,104 +1,478 @@
+    if (event.createItem) {
+      const ci = event.createItem;
+      const id = bytesToHex(ci.eventId);
+      const item = {
+        id,
+        price: ci.price,
+        metadata: Uint8ArrayToParsedJSON(ci.metadata),
+        tags: [],
+        quantity: 0,
+      };
+      await this.store.put(id, item);
+    } else if (event.updateItem) {
+      const ui = event.updateItem;
+      const id = bytesToHex(ui.itemId);
+      const item = await this.store.get(id);
+      if (ui.metadata) {
+        item.metadata = Uint8ArrayToParsedJSON(ui.metadata);
+      }
+      if (ui.price) {
+        item.price = ui.price;
+      }
+      await this.store.put(id, item);
+    } else if (event.changeStock) {
+      const cs = event.changeStock;
+      if (cs.itemIds) {
+        cs.itemIds.forEach(async (id: Uint8Array, i: number) => {
+          const itemId = bytesToHex(id);
+          const item = await this.store.get(itemId);
+          const diff = cs.diffs ? cs.diffs[i] : 0;
+          item.quantity = item.quantity + diff;
+          await this.store.put(itemId, item);
+        });
+      }
+    } else if (event.updateTag) {
+      // Add or remove tagId to item
+      const ut = event.updateTag;
+      const tagId = bytesToHex(ut.tagId);
+
+      if (ut.addItemId) {
+        const itemId = bytesToHex(ut.addItemId);
+        const item = await this.store.get(itemId);
+        item.tags = [...item.tags, tagId];
+        await this.store.put(itemId, item);
+      }
+      if (ut.removeItemId) {
+        const itemId = bytesToHex(ut.removeItemId);
+        const item = await this.store.get(itemId);
+        item.tags = [...item.tags.filter((id: `0x${string}`) => id !== tagId)];
+        await this.store.put(itemId, item);
+      }
+    }
+  }
+
+  async create(item: Item) {
+    const id = bytesToHex(
+      await this.client.createItem({
+        price: item.price,
+        metadata: new TextEncoder().encode(JSON.stringify(item.metadata)),
+      }),
+    );
+    await this.store.put(id, item);
+  }
+  //update argument passed here will only contain the fields to update.
+  async update(update: Partial<Item>) {
+    //ui object to be passed to the network with converted network data types
+    const ui: schema.IUpdateItem = {
+      itemId: hexToBytes(update.id!),
+    };
+    if (update.price) {
+      ui.price = update.price;
+    }
+    if (update.metadata) {
+      ui.metadata = new TextEncoder().encode(JSON.stringify(update.metadata));
+    }
+    const p1 = this.client.updateItem(ui);
+    const p2 = async () => {
+      //Get the full item, update the new fields, and save to listings store
+      const item = await this.store.get(update.id!);
+      if (update.price) {
+        item.price = update.price;
+      }
+      if (update.metadata) {
+        item.metadata = update.metadata;
+      }
+      if (update.quantity) {
+        const diff = update.quantity - item.quantity;
+        await this.client.changeStock({
```

Comment:
This should be pulled out of the current closure and pushed to the opQeueu
```suggestion
       opQueue.push(this.client.changeStock({
```

Reply:
i think the changestock promise has to be a part of whatever closure `const item = await this.store.get(update.id!);` is in because we have to calculate the diff